src/wasm/kotlin/src/arcs/Particle.kt

```diff
@@ -10,55 +10,143 @@ abstract class Particle : WasmObject() {
+     * @param handle Singleton or Collection, defined in this particle class
+     */
     fun registerHandle(name: String, handle: Handle) {
         handle.name = name
         handle.particle = this
         handles[name] = handle
         log("Registering $name")
     }
 
-    fun eventHandler(name: String, handler: (Map<String, String>) -> Unit) {
-        eventHandlers[name] = handler
-    }
-
-    fun connectHandle(handleName: String, willSync: Boolean): Handle? {
-        log("Connect called internal '$handleName'")
+    /**
+     * Connect to a registered handle
+     *
+     * If a handle has been previously registered, return the handle. Optionally, mark the handle for later
+     * synchronization.
+     *
+     * @param name Name of the handle
+     * @param willSync Mark handle for synchronization; depends if handle is readable
+     * @return The name-associated handle, or null
+     * @see [registerHandle]
+     * @see [onHandleSync]
+     */
+    fun connectHandle(name: String, willSync: Boolean): Handle? {
+        log("Connect called internal '$name'")
 
-        handles[handleName]?.let {
+        handles[name]?.let {
             if (willSync) toSync.add(it)
             return it
         }
 
-        log("Handle $handleName not registered")
+        log("Handle $name not registered")
         return null
     }
 
+    /**
+     * Register a reaction to an event.
+     *
+     * Particle templates may emit events, usually from user actions.
+     *
+     * @param name The name of the triggered event
+     * @param handler A callback (consumer) in reaction to the event
+     */
+    fun eventHandler(name: String, handler: (Map<String, String>) -> Unit) {
+        eventHandlers[name] = handler
+    }
+
+    /**
+     * Trigger an event.
+     *
+     * Will target registered events, if present. Will always initiate rendering when called.
+     *
+     * @param slotName Slot that the event is associated with; likely `root`
+     * @param eventName Name of the event to trigger
+     * @param eventData Data associated with the event; will be passed into event handler
+     * @see [eventHandler]
+     */
+    open fun fireEvent(slotName: String, eventName: String, eventData: Map<String, String>) {
+        eventHandlers[eventName]?.invoke(eventData)
+        renderOutput()
+    }
+
+    /** @param handle Handle to synchronize */
     fun sync(handle: Handle) {
         log("Particle.sync called")
         toSync.remove(handle)
         onHandleSync(handle, toSync.isEmpty())
     }
 
-    open fun onHandleUpdate(handle: Handle) {}
-    open fun onHandleSync(handle: Handle, allSynced: Boolean) {}
 
+    /**
+     * React to handle updates.
+     *
+     * Called for handles when change events are received from the backing store. Default action is to trigger
+     * rendering.
+     *
+     * @param handle Singleton or Collection handle
+     */
+    open fun onHandleUpdate(handle: Handle) = renderOutput()
```

Comment:
I'm not sure we want auto rendering to be the default behaviour. In C++ I have this as an option but it requires explicit enabling by the particle dev. I don't think we should make too many assumptions about what type of particles are going to be most common, so IMO it's better to have convenience features that must be opted into, rather than out of.

Reply:
I'm agree with Myk. The so-called `SimpleParticle` has internal code that buffers handle-changes and invokes the rendering at a maximum frequency. IMO for closer-to-the-metal objects (aka `Particle.js`) it's the author's responsibility to invoke rendering when appropriate (which could be directly onHandleUpdate).

Reply:
Good shout. Reverting 8a80159